--- Day 3: Gear Ratios ---
You and the Elf eventually reach a gondola lift station; he says the gondola lift will take you up to the water source, but this is as far as he can bring you. You go inside.

It doesn't take long to find the gondolas, but there seems to be a problem: they're not moving.

"Aaah!"

You turn around to see a slightly-greasy Elf with a wrench and a look of surprise. "Sorry, I wasn't expecting anyone! The gondola lift isn't working right now; it'll still be a while before I can fix it." You offer to help.

The engineer explains that an engine part seems to be missing from the engine, but nobody can figure out which one. If you can add up all the part numbers in the engine schematic, it should be easy to work out which part is missing.

The engine schematic (your puzzle input) consists of a visual representation of the engine. There are lots of numbers and symbols you don't really understand, but apparently any number adjacent to a symbol, even diagonally, is a "part number" and should be included in your sum. (Periods (.) do not count as a symbol.)

Here is an example engine schematic:

467..114..
...*......
..35..633.
......#...
617*......
.....+.58.
..592.....
......755.
...$.*....
.664.598..
In this schematic, two numbers are not part numbers because they are not adjacent to a symbol: 114 (top right) and 58 (middle right). Every other number is adjacent to a symbol and so is a part number; their sum is 4361.

Of course, the actual engine schematic is much larger. What is the sum of all of the part numbers in the engine schematic?



In [40]:
import re
with open('input03.txt') as file:
    data = [line.rstrip() for line in file]
data[0]

'...733.......289..262.....520..................161.462..........450.........................183.............................................'

In [5]:
sym_uniques = set()
for line in data:
    for symbol in re.findall(r'[^\d\.]', line):
        sym_uniques.add(symbol)
sym_uniques

{'#', '$', '%', '&', '*', '+', '-', '/', '=', '@'}

In [6]:
sym_pos = set()
for i, line in enumerate(data):
    for j, char in enumerate(line):
        if char in sym_uniques:
            sym_pos.add((i, j))
sym_pos

{(1, 4),
 (1, 25),
 (1, 50),
 (1, 63),
 (1, 91),
 (2, 39),
 (2, 46),
 (2, 83),
 (2, 106),
 (2, 121),
 (2, 124),
 (2, 129),
 (2, 136),
 (3, 17),
 (3, 27),
 (3, 56),
 (3, 77),
 (3, 94),
 (4, 10),
 (4, 33),
 (4, 43),
 (4, 72),
 (4, 125),
 (5, 12),
 (5, 51),
 (6, 3),
 (6, 17),
 (6, 24),
 (6, 35),
 (6, 58),
 (6, 89),
 (6, 100),
 (6, 114),
 (6, 120),
 (7, 41),
 (7, 49),
 (7, 68),
 (7, 136),
 (8, 22),
 (8, 105),
 (8, 109),
 (8, 129),
 (9, 6),
 (9, 49),
 (9, 52),
 (9, 58),
 (9, 65),
 (9, 77),
 (9, 82),
 (9, 98),
 (9, 118),
 (10, 4),
 (10, 27),
 (10, 34),
 (10, 44),
 (10, 71),
 (10, 88),
 (10, 125),
 (11, 18),
 (11, 24),
 (11, 57),
 (11, 74),
 (11, 104),
 (11, 109),
 (11, 123),
 (12, 61),
 (12, 95),
 (12, 128),
 (13, 12),
 (13, 34),
 (13, 121),
 (14, 6),
 (14, 26),
 (14, 43),
 (14, 55),
 (14, 85),
 (14, 102),
 (14, 107),
 (14, 113),
 (14, 129),
 (15, 4),
 (15, 38),
 (15, 79),
 (15, 91),
 (16, 21),
 (16, 53),
 (16, 63),
 (16, 67),
 (16, 76),
 (16, 104),
 (16, 110),
 (17, 4),
 (17, 44),
 (17, 133

In [25]:
numbers = []
for i, line in enumerate(data):
    for m in re.finditer(r'\d+', line):
        numbers.append((i, m.start(), m.end() - 1, int(m[0])))
numbers

[(0, 3, 5, 733),
 (0, 13, 15, 289),
 (0, 18, 20, 262),
 (0, 26, 28, 520),
 (0, 47, 49, 161),
 (0, 51, 53, 462),
 (0, 64, 66, 450),
 (0, 92, 94, 183),
 (1, 39, 41, 707),
 (1, 43, 45, 352),
 (1, 85, 87, 801),
 (1, 107, 109, 333),
 (1, 112, 114, 196),
 (1, 123, 125, 484),
 (1, 127, 129, 635),
 (1, 136, 138, 287),
 (2, 4, 5, 42),
 (2, 15, 17, 131),
 (2, 22, 24, 913),
 (2, 57, 59, 634),
 (2, 78, 80, 440),
 (2, 99, 100, 83),
 (2, 118, 120, 404),
 (2, 132, 134, 423),
 (3, 0, 2, 618),
 (3, 10, 12, 272),
 (3, 34, 36, 547),
 (3, 38, 40, 344),
 (3, 69, 71, 689),
 (3, 73, 75, 589),
 (3, 82, 84, 150),
 (3, 91, 93, 382),
 (3, 127, 129, 168),
 (3, 136, 138, 433),
 (4, 26, 28, 253),
 (4, 30, 32, 102),
 (4, 50, 51, 78),
 (4, 59, 61, 804),
 (4, 97, 99, 858),
 (5, 3, 4, 69),
 (5, 13, 14, 37),
 (5, 18, 20, 510),
 (5, 22, 24, 797),
 (5, 36, 38, 596),
 (5, 40, 42, 946),
 (5, 86, 88, 602),
 (5, 90, 92, 175),
 (5, 108, 110, 203),
 (5, 113, 115, 100),
 (5, 126, 128, 681),
 (5, 136, 138, 546),
 (6, 9, 11, 110),

In [30]:
def is_part_num(number, sym_pos):
    n_i, n_j_s, n_j_e, _ = number
    for s_i, s_j in sym_pos:
        if n_i - 1 <= s_i <= n_i + 1:
            if n_j_s - 1 <= s_j <= n_j_e + 1:
                return True
    return False
is_part_num((0,0,0,0), sym_pos)

False

In [31]:
sum_part_num = 0
for num in numbers:
    if is_part_num(num, sym_pos):
        sum_part_num += num[3]
sum_part_num

525181

Your puzzle answer was 525181.

The first half of this puzzle is complete! It provides one gold star: *

--- Part Two ---
The engineer finds the missing part and installs it in the engine! As the engine springs to life, you jump in the closest gondola, finally ready to ascend to the water source.

You don't seem to be going very fast, though. Maybe something is still wrong? Fortunately, the gondola has a phone labeled "help", so you pick it up and the engineer answers.

Before you can explain the situation, she suggests that you look out the window. There stands the engineer, holding a phone in one hand and waving with the other. You're going so slowly that you haven't even left the station. You exit the gondola.

The missing part wasn't the only issue - one of the gears in the engine is wrong. A gear is any * symbol that is adjacent to exactly two part numbers. Its gear ratio is the result of multiplying those two numbers together.

This time, you need to find the gear ratio of every gear and add them all up so that the engineer can figure out which gear needs to be replaced.

Consider the same engine schematic again:

467..114..
...*......
..35..633.
......#...
617*......
.....+.58.
..592.....
......755.
...$.*....
.664.598..
In this schematic, there are two gears. The first is in the top left; it has part numbers 467 and 35, so its gear ratio is 16345. The second gear is in the lower right; its gear ratio is 451490. (The * adjacent to 617 is not a gear because it is only adjacent to one part number.) Adding up all of the gear ratios produces 467835.

What is the sum of all of the gear ratios in your engine schematic?

In [37]:
stars = {}
for i, line in enumerate(data):
    for j, char in enumerate(line):
        if char == '*':
            stars[i, j] = []
stars[(72, 6)]

[]

In [38]:
for num in numbers:
    n_i, n_j_s, n_j_e, n_val = num
    for s_i, s_j in stars:
        if n_i - 1 <= s_i <= n_i + 1:
            if n_j_s - 1 <= s_j <= n_j_e + 1:
                stars[(s_i, s_j)].append(n_val)
stars

{(1, 4): [733, 42],
 (1, 25): [520, 913],
 (1, 50): [161, 462],
 (2, 39): [707, 344],
 (2, 129): [635, 168],
 (2, 136): [287, 433],
 (3, 17): [131],
 (3, 77): [440],
 (4, 33): [547, 102],
 (4, 72): [689, 589],
 (5, 12): [37, 110],
 (6, 3): [69, 973],
 (6, 17): [510, 274],
 (6, 24): [797],
 (6, 35): [596, 576],
 (6, 89): [602, 175],
 (6, 100): [246, 637],
 (7, 68): [381, 482],
 (7, 136): [222],
 (8, 129): [883, 498],
 (9, 6): [223],
 (9, 58): [701, 936],
 (9, 65): [17, 24],
 (9, 82): [80, 39],
 (9, 118): [563, 275],
 (10, 4): [373, 974],
 (10, 34): [707, 965],
 (10, 44): [956, 450],
 (10, 88): [698, 437],
 (11, 18): [952, 308],
 (11, 57): [540, 4],
 (11, 109): [514, 670],
 (12, 95): [186, 58],
 (13, 12): [58, 587],
 (13, 34): [34, 599],
 (13, 121): [975, 960],
 (14, 6): [22, 406],
 (14, 43): [880, 6],
 (14, 55): [939, 326],
 (14, 129): [346],
 (15, 79): [831, 611],
 (15, 91): [552, 79],
 (16, 21): [828],
 (16, 53): [764],
 (16, 67): [722, 932],
 (16, 76): [334, 590],
 (16, 110): [546, 9

In [39]:
sum_ratios = 0
for parts in stars.values():
    if len(parts) == 2:
        sum_ratios += parts[0] * parts[1]
sum_ratios

84289137

That's the right answer! You are one gold star closer to restoring snow operations.